### ISSUE TITLE
Fatal error: Problematic frame:  V  [libjvm.dylib+0x6d20a6]  SymbolTable::lookup_dynamic(int, char const*, int, unsigned int)+0x1c

### ISSUE THREAD

**Issue by oncecrossover** [brevilo/jolm#4]
Hey @brevilo , sometimes, It has an fatal error on unit test, it will disappear after many runs.

there is log message on Mac OS 11.5
```
/Library/Java/JavaVirtualMachines/jdk-11.0.1.jdk/Contents/Home/bin/java -ea -Didea.test.cyclic.buffer.size=1048576 "-javaagent:/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=58487:/Applications/IntelliJ IDEA.app/Contents/bin" -Dfile.encoding=UTF-8 -classpath "/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/junit/lib/junit-rt.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/junit/lib/junit5-rt.jar:/Users/Wendell/.m2/repository/org/junit/platform/junit-platform-launcher/1.7.2/junit-platform-launcher-1.7.2.jar:/Users/Wendell/.m2/repository/org/apiguardian/apiguardian-api/1.1.0/apiguardian-api-1.1.0.jar:/Users/Wendell/.m2/repository/org/junit/platform/junit-platform-engine/1.7.2/junit-platform-engine-1.7.2.jar:/Users/Wendell/.m2/repository/org/opentest4j/opentest4j/1.2.0/opentest4j-1.2.0.jar:/Users/Wendell/.m2/repository/org/junit/platform/junit-platform-commons/1.7.2/junit-platform-commons-1.7.2.jar:/Users/Wendell/workspace/jolm/target/test-classes:/Users/Wendell/workspace/jolm/target/classes:/Users/Wendell/.m2/repository/net/java/dev/jna/jna/5.8.0/jna-5.8.0.jar:/Users/Wendell/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.11.1/jackson-databind-2.11.1.jar:/Users/Wendell/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.11.1/jackson-annotations-2.11.1.jar:/Users/Wendell/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.11.1/jackson-core-2.11.1.jar:/Users/Wendell/.m2/repository/org/junit/jupiter/junit-jupiter-engine/5.7.2/junit-jupiter-engine-5.7.2.jar:/Users/Wendell/.m2/repository/org/junit/jupiter/junit-jupiter-api/5.7.2/junit-jupiter-api-5.7.2.jar" com.intellij.rt.execution.junit.JUnitStarter -ideVersion5 -junit5 io.github.brevilo.jolm.SessionTest,testAliceToBob
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000010e6d20a6, pid=15556, tid=6147
#
# JRE version: OpenJDK Runtime Environment (11.0.1+13) (build 11.0.1+13)
# Java VM: OpenJDK 64-Bit Server VM (11.0.1+13, mixed mode, tiered, compressed oops, g1 gc, bsd-amd64)
# Problematic frame:
# V  [libjvm.dylib+0x6d20a6]  SymbolTable::lookup_dynamic(int, char const*, int, unsigned int)+0x1c
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/Wendell/workspace/jolm/hs_err_pid15556.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#

Process finished with exit code 134 (interrupted by signal 6: SIGABRT)

[hs_err_pid15556.log](https://github.com/brevilo/jolm/files/7524446/hs_err_pid15556.log)

```

**Comment by brevilo** [comment]
This is a strange one. While I can reproduce it occasionally in Eclipse, ~~I never can reproduce it via Maven~~. Where do you encounter it?

**Comment by brevilo** [comment]
For reference: the JVM segfault is triggered while deserializing the identity keys JSON string into the `IdentityKeys` POJO in [`Account.identityKeys()`](https://github.com/brevilo/jolm/blob/d50eb8140d994f2e62cb25ac9acf0f11993efca1/src/main/java/io/github/brevilo/jolm/Account.java#L80) using FasterXML's jackson-databind.

No idea why. It also happens with a fixed simple JSON literal. Seems to be related to class loading (presumably of the POJO) and I see references to `java/security/ProtectionDomain`. Hm...

**Comment by brevilo** [comment]
Sooo.... this one turned out to be a tricky little bugger! Should be fixed now in branch `fix-4`.

@oncecrossover, please give it a try and report back. I'll merge #6  if it fixes the issue for you as well.

Note: there's a reason why I list "Memory management audit" as part of the TODOs in the [README](https://github.com/brevilo/jolm/blob/main/README.md#status) 😏

**Comment by brevilo** [comment]
I chose to merge the fix without your (additional) test such that `main` gets fixed...

**Comment by oncecrossover** [comment]
Sorry! I forgot this ticket.

I have tested it many times but still exists. Remarkably, it only seems to happen in unit tests.

**Comment by oncecrossover** [comment]
```
git diff:

-@RunWith(SpringJUnit4ClassRunner.class)
-@SpringBootTest(classes = TestApplication.class)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
```

I got it, this error happens with ```SpringJUnit4ClassRunner/SpringBootTest```.  It works fine with ```TestInstance```, and never happens with ```TestInstance```.
But why?

May be the same problem in  #7.

**Comment by brevilo** [comment]
I can't reproduce the issue. Please provide a test case that isolates the issue and allows me to reproduce it.

Also, which version of jOlm and which JVM are you using now? Is the error's stacktrace still the same?

**Comment by brevilo** [comment]
> I got it, this error happens with `SpringJUnit4ClassRunner/SpringBootTest`. It works fine with `TestInstance`, and never happens with `TestInstance`.
> But why?

Here's an idea: [`TestInstance`](https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/TestInstance.html) causes all tests methods to share their state by using the `PER_CLASS` lifecycle (instead of the `PER_METHOD` default). Since you don't provide your actual test code I have to assume that your Spring-based unit tests might invalidate the state and in turn the memory blocks backing the `Account` and/or `Session` instances between the individual test methods. Unless you can prove this to be wrong (by a test case reproducing the issue), I think the error you see is an artifact of the way you run *your* unit tests.

**Comment by oncecrossover** [comment]
TestOlmSession.java
```
package net.metachain.server.utils;

import io.github.brevilo.jolm.Account;
import io.github.brevilo.jolm.Session;
import io.github.brevilo.jolm.model.Message;
import net.metachain.server.TestApplication;
import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;

import java.util.Map;

@RunWith(SpringJUnit4ClassRunner.class)
@SpringBootTest(classes = TestApplication.class)
public class TestOlmSession {

    private static final Logger LOGGER = LoggerFactory.getLogger(TestOlmSession.class);

    /**
     * Basic test:
     * - alice creates an account
     * - bob creates an account
     * - alice creates an outbound session with bob (bobIdentityKey & bobOneTimeKey)
     * - alice encrypts a message with its session
     * - bob creates an inbound session based on alice's encrypted message
     * - bob decrypts the encrypted message with its session
     */
    @Test
    public void testAliceToBob() {
        String plainMessage = "Hello Bob, this is message from Alice.";
        LOGGER.info("plain message: {}", plainMessage);
        String decryptedMessage = "";

        String bobIdentityKey = null;
        String bobOneTimeKey = null;

        Account aliceAccount = null;
        Account bobAccount = null;

        Session aliceSession = null;
        Session bobSession = null;

        try {
            // ALICE & BOB ACCOUNTS CREATION
            aliceAccount = new Account();
            Assert.assertNotNull(aliceAccount);

            bobAccount = new Account();
            Assert.assertNotNull(bobAccount);

            // generate identityKey
            bobIdentityKey = bobAccount.identityKeys().getCurve25519();

            /**
             * generate oneTimeKey
             */
            bobAccount.generateOneTimeKeys(5);
            Map<String, String> bobOneTimeKeys = bobAccount.oneTimeKeys().getCurve25519();
            for (String key : bobOneTimeKeys.keySet()) {
                bobOneTimeKey = bobOneTimeKeys.get(key);
                LOGGER.info("OneTimeKeys: {} = {}", key, bobOneTimeKey);
            }
            LOGGER.info("identityKey = {}, oneTimeKey = {}", bobIdentityKey, bobOneTimeKey);

            // create OutboundSession to Bob
            aliceSession = Session.createOutboundSession(aliceAccount, bobIdentityKey, bobOneTimeKey);
            Assert.assertNotNull(aliceSession);

            Message encryptedMessage = aliceSession.encrypt(plainMessage);
            LOGGER.info("encrypted message: {}", JSONUtils.toJSONString(encryptedMessage));
            LOGGER.info("encrypted message type: {}", encryptedMessage.type());

            bobSession = Session.createInboundSession(bobAccount, encryptedMessage.getCipherText());
            Assert.assertNotNull(bobSession);

            decryptedMessage = bobSession.decrypt(encryptedMessage);

            LOGGER.info("decrypted message: {}", decryptedMessage);
        } catch (Exception e) {
            LOGGER.error("Olm session test error: ", e);
        } finally {
            aliceAccount.clear();
            bobAccount.clear();

            aliceSession.clear();
            bobSession.clear();
        }
        Assert.assertEquals(plainMessage, decryptedMessage);
    }

}

```


TestApplication.java
```
@SpringBootApplication(scanBasePackages = {"net.metachain.server"},
        exclude = {DataSourceAutoConfiguration.class})
public class TestApplication {

    public static void main(String[] args) {
        SpringApplication.run(TestApplication.class);
    }

}

```

This is my full code, I used spring-boot framework, you can use it to reproduce the error. run three or five times you will reproduce it.

**Comment by oncecrossover** [comment]
```
package net.metachain.server.utils;

import io.github.brevilo.jolm.Account;
import io.github.brevilo.jolm.Session;
import io.github.brevilo.jolm.model.Message;
import org.junit.Assert;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;

public class TestOlmSession {

    private static final Logger LOGGER = LoggerFactory.getLogger(TestOlmSession.class);

    /**
     * Basic test:
     * - alice creates an account
     * - bob creates an account
     * - alice creates an outbound session with bob (bobIdentityKey & bobOneTimeKey)
     * - alice encrypts a message with its session
     * - bob creates an inbound session based on alice's encrypted message
     * - bob decrypts the encrypted message with its session
     */
    @Test
    public void testAliceToBob() {
        String plainMessage = "Hello Bob, this is message from Alice.";
        LOGGER.info("plain message: {}", plainMessage);
        String decryptedMessage = "";

        String bobIdentityKey = null;
        String bobOneTimeKey = null;

        Account aliceAccount = null;
        Account bobAccount = null;

        Session aliceSession = null;
        Session bobSession = null;

        try {
            // ALICE & BOB ACCOUNTS CREATION
            aliceAccount = new Account();
            Assert.assertNotNull(aliceAccount);

            bobAccount = new Account();
            Assert.assertNotNull(bobAccount);

            // generate identityKey
            bobIdentityKey = bobAccount.identityKeys().getCurve25519();

            /**
             * generate oneTimeKey
             */
            bobAccount.generateOneTimeKeys(5);
            Map<String, String> bobOneTimeKeys = bobAccount.oneTimeKeys().getCurve25519();
            for (String key : bobOneTimeKeys.keySet()) {
                bobOneTimeKey = bobOneTimeKeys.get(key);
                LOGGER.info("OneTimeKeys: {} = {}", key, bobOneTimeKey);
            }
            LOGGER.info("identityKey = {}, oneTimeKey = {}", bobIdentityKey, bobOneTimeKey);

            // create OutboundSession to Bob
            aliceSession = Session.createOutboundSession(aliceAccount, bobIdentityKey, bobOneTimeKey);
            Assert.assertNotNull(aliceSession);

            Message encryptedMessage = aliceSession.encrypt(plainMessage);
            LOGGER.info("encrypted message: {}", JSONUtils.toJSONString(encryptedMessage));
            LOGGER.info("encrypted message type: {}", encryptedMessage.type());

            bobSession = Session.createInboundSession(bobAccount, encryptedMessage.getCipherText());
            Assert.assertNotNull(bobSession);

            decryptedMessage = bobSession.decrypt(encryptedMessage);

            LOGGER.info("decrypted message: {}", decryptedMessage);
        } catch (Exception e) {
            LOGGER.error("Olm session test error: ", e);
        } finally {
            aliceAccount.clear();
            bobAccount.clear();

            aliceSession.clear();
            bobSession.clear();
        }
        Assert.assertEquals(plainMessage, decryptedMessage);
    }

}

```

This is my new test code, not used spring-boot test framework, it never reproduce the error.

**Comment by brevilo** [comment]
Thanks, I'm going to look into the issue but with a somewhat lower priority since, so far, it concerns only a *specific* unit test framework you use *and* there's a workaround for it.

Stay tuned...

**Comment by brevilo** [comment]
So, here we go again.

Please check out branch `fix-4-2` and see if it fixes this issue (as well as #7) for you and let me know.

**Comment by oncecrossover** [comment]
Hey @brevilo , I have tested with branch fix-4-2, It looks great. then we will deploy to live environment to validate it.

Please wait...

**Comment by brevilo** [comment]
That's good news! Please try and make sure to throw all tests you have at it (incl. those that triggered #7). Would be great if we could wrap this up once and for all 😏

**Comment by brevilo** [comment]
Hey @oncecrossover, any updates on your tests? I'd like to merge the fixes as soon as possible so that everyone can benefit from them. There are also a few other changes/improvements in the pipeline waiting for the next release.

Thanks

**Comment by oncecrossover** [comment]
Sorry @brevilo, I had planned to test it for a week.
But now, let me tell you good news, our service is running start from  ```UTC 2021-11-21 19:04:36```, it works fine! Great job! 👍

**Comment by oncecrossover** [comment]
I think this bug has been fixed, you can release a new version.

**Comment by brevilo** [comment]
Ok, thanks. I'll merge the fix and close this one. A new release should be published over the weekend (at the latest).
